### PR TITLE
Enable more linting rules

### PR DESCRIPTION
**Pull request by g0t00**

### CODE REVIEW

File: lib/parser/interfaces.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export interface IMayHaveLabel {
   labelLinks: OLabelReference[];
 }
 export function implementsIHasLabel(obj: ObjectBase): obj is ObjectBase & IHasLabel {
-  return (obj as ObjectBase & IHasLabel).label !== undefined && Array.isArray((obj as ObjectBase & IHasLabel).labelLinks);
+  return (obj as ObjectBase & Partial<IHasLabel>).label !== undefined && Array.isArray((obj as ObjectBase & IHasLabel).labelLinks);
```
Comment (Nik-Sch) [inline lib/parser/interfaces.ts:15]:
```suggestion
  return (obj as ObjectBase & IHasLabel).label !== undefined && Array.isArray((obj as ObjectBase & IHasLabel).labelLinks);
```
partial is not required here and in the whole file. Undo these changes

File: lib/parser/objects.ts [hunk 2]
```diff
@@ -30,7 +27,10 @@ export class OI implements Position {
     if (this.i_ === undefined) {
       this.calcI();
     }
-    return this.i_ as number;
+    if (this.i_ === undefined) {
+      throw new Error('Can not convert position to i');
+    }
```
Comment (Nik-Sch) [inline lib/parser/objects.ts:32]: maybe move the check in the if condition above. Otherwise this looks a bit weird why it is checked twice...

File: lib/parser/objects.ts [hunk 3]
```diff
@@ -3,13 +3,10 @@ import { OLexerToken } from '../lexer';
...
 export class OI implements Position {
   protected i_?: number;
-  constructor(parent: ObjectBase | OFile, i: number)
-  constructor(parent: ObjectBase | OFile, line: number, character: number)
-  constructor(parent: ObjectBase | OFile, line: number, character: number, i: number)
   constructor(public parent: ObjectBase | OFile, i: number, j?: number, k?: number) {
     if (j === Number.POSITIVE_INFINITY) {
       const lines = (this.parent instanceof OFile ? this.parent : this.parent.rootFile).lines;
-      j = lines[i].length - 1;
+      j = lines[i]!.length - 1;
```
Comment (Nik-Sch) [inline lib/parser/objects.ts:9]: `lines[i]` might actually be undefined, should be checked

File: lib/parser/parser-base.ts [hunk 4]
```diff
@@ -9,10 +10,12 @@ export class ParserPosition {
   public file: OFile;
   public num = 0;
   public get i() {
-    if (this.num >= this.lexerTokens.length) {
-      throw new ParserError(`I out of range`, this.lexerTokens[this.lexerTokens.length - 1].range);
+    const token = this.lexerTokens[this.num];
+    if (!token) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:15]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 5]
```diff
@@ -21,7 +24,8 @@ export class ParserPosition {
     return this.num >= 0 && this.num < this.lexerTokens.length;
   }
   public getRangeToEndLine() {
-    return this.lexerTokens[this.num]?.range.copyExtendEndOfLine();
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:27]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 6]
```diff
@@ -37,58 +41,69 @@ export class ParserBase {
...
   }
 
 
   // Offset gives an offset to the current parser position. If offsetIgnoresWhitespaces is set whitespace (and comment) is not counted.
   // Meaning offset = 2 counts only the next two non-whitespaces tokens
-  getToken(offset = 0, offsetIgnoresWhitespaces = false) {
+  getToken(offset = 0, offsetIgnoresWhitespaces = false): OLexerToken {
     if (!this.state.pos.isValid()) {
-      throw new ParserError(`EOF reached`, this.state.pos.lexerTokens[this.state.pos.lexerTokens.length - 1].range);
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:53]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 7]
```diff
@@ -37,58 +41,69 @@ export class ParserBase {
...
-              throw new ParserError(`Out of bound while doing getToken(${offset}, ${offsetIgnoresWhitespaces})`, this.getToken(0).range);
+            if (this.state.pos.num + offsetCorrected >= this.state.pos.lexerTokens.length) {
+              throw new ParserError(`Out of bound while doing getToken(${offset}, ${String(offsetIgnoresWhitespaces)})`, this.getToken(0).range);
             }
-          } while ((this.state.pos.lexerTokens[this.state.pos.num + offsetCorrected].isWhitespace()));
+          } while ((this.state.pos.lexerTokens[this.state.pos.num + offsetCorrected]?.isWhitespace()));
         }
       } else if (offset === 0) {
-        return this.state.pos.lexerTokens[this.state.pos.num];
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:77]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 8]
```diff
@@ -37,58 +41,69 @@ export class ParserBase {
...
   }
-  consumeToken(advanceWhitespace = true) {
+  consumeToken(advanceWhitespace = true): OLexerToken {
     const token = this.state.pos.lexerTokens[this.state.pos.num];
     this.state.pos.num++;
     if (advanceWhitespace) { // This should not be necessary anymore, if everything is correctly using tokens
       this.advanceWhitespace();
     }
+    if (!token) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:101]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 9]
```diff
@@ -112,10 +127,11 @@ export class ParserBase {
       }
       return false;
     }
-    while (checkToken(this.state.pos.lexerTokens[this.state.pos.num]) === false) {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:130]:
```suggestion
```
has been globaly disabled

File: lib/languageFeatures/signatureHelp.ts [hunk 10]
```diff
@@ -51,10 +51,11 @@ export function signatureHelp(linter: VhdlLinter, position: Position): Signature
             for (const [childNumber, child] of associationList.children.entries()) {
               // Extend the end range by white spaces (assume it belongs to the association if cursor is in whitespace)
               let tokenIndex = linter.file.lexerTokens.findIndex(token => token.range.end.i === child.range.end.i);
-              while (linter.file.lexerTokens[tokenIndex + 1].isWhitespace()) {
+              while (linter.file.lexerTokens[tokenIndex + 1]?.isWhitespace()) {
                 tokenIndex++;
               }
-              if (posI >= linter.file.lexerTokens[tokenIndex].range.end.i) {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/languageFeatures/signatureHelp.ts:57]:
```suggestion
```
has been globaly disabled

File: lib/parser/alias-parser.ts [hunk 11]
```diff
@@ -90,14 +90,20 @@ export class AliasParser extends ParserBase {
...
+    if (!firstToken) {
+      throw new ParserError(`expected token`, parent.range);
+    }
+    if (secondToken) {
+      const prefix = [new OReference(parent, firstToken)];
       for (const token of tokens.slice(1)) {
         prefix.push(new OSelectedName(parent, token, prefix.slice() as SelectedNamePrefix));
       }
-      return prefix[prefix.length - 1];
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/alias-parser.ts:104]:
```suggestion
```
has been globaly disabled

File: lib/parser/assignment-parser.ts [hunk 12]
```diff
@@ -18,7 +18,8 @@ export class AssignmentParser extends ParserBase {
     this.findToken(['<=', ':=']);
     const leftHandSideTokens = [];
     while (leftHandSideNum < this.state.pos.num) {
-      leftHandSideTokens.push(this.state.pos.lexerTokens[leftHandSideNum]);
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/assignment-parser.ts:21]:
```suggestion
```
has been globaly disabled

File: lib/parser/association-list-parser.ts [hunk 13]
```diff
@@ -20,7 +20,8 @@ export class AssociationListParser extends ParserBase {
       // eslint-disable-next-line prefer-const
       let [associationTokens, lastChar] = this.advanceParenthesisAware([',', ')']);
       if (associationTokens.length > 0) {
-        const combinedRange = associationTokens[0].range.copyWithNewEnd(associationTokens[associationTokens.length - 1].range);
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/association-list-parser.ts:23]:
```suggestion
```
has been globaly disabled

File: lib/parser/expression-parser.ts [hunk 14]
```diff
@@ -173,7 +177,8 @@ export class ExpressionParser extends ParserBase {
     if (this.expState.lastFormal.length > 0 && tokenBuffer.length === 0) {
       this.state.messages.push({
         message: "The actual part cannot be empty.",
-        range: this.expState.lastFormal[0].range.copyWithNewEnd(this.expState.lastFormal[this.expState.lastFormal.length - 1].range)
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/expression-parser.ts:180]:
```suggestion
```
has been globaly disabled

File: lib/parser/objects.ts [hunk 15]
```diff
@@ -66,15 +66,16 @@ export class OI implements Position {
   private calcPosition(): Position {
     const lines = (this.parent instanceof OFile ? this.parent : this.parent.rootFile).text.slice(0, this.i).split('\n');
     const line = lines.length - 1;
-    const character = lines[lines.length - 1].length;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/objects.ts:69]:
```suggestion
```
has been globaly disabled

File: lib/parser/subtype-parser.ts [hunk 16]
```diff
@@ -23,7 +23,8 @@ export class SubtypeParser extends ParserBase {
     this.subtype.range = this.subtype.range.copyWithNewEnd(this.state.pos.i);
     const tokens = this.advanceSemicolon(true);
     if (tokens.length > 0) {
-      this.subtype.range = this.subtype.range.copyWithNewEnd(tokens[tokens.length - 1].range.end.i);
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/parser/subtype-parser.ts:26]:
```suggestion
```
has been globaly disabled

File: lib/rules/instantiations.ts [hunk 17]
```diff
@@ -40,15 +40,17 @@ export class RInstantiation extends RuleBase implements IRule {
...
+          if (possibleMatches.length > 0 && firstFormal) {
+            const bestMatch = findBestMatch(firstFormal.referenceToken.text, possibleMatches);
             code = this.vhdlLinter.addCodeActionCallback((textDocumentUri: string) => {
               const actions = [];
               actions.push(CodeAction.create(
                 `Replace with ${bestMatch.bestMatch.target} (score: ${bestMatch.bestMatch.rating})`,
                 {
                   changes: {
-                    [textDocumentUri]: [TextEdit.replace(Range.create(association.formalPart[0].range.start, association.formalPart[association.formalPart.length - 1].range.end)
+                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (Nik-Sch) [inline lib/rules/instantiations.ts:52]:
```suggestion
```
has been globaly disabled

File: lib/parser/parser-base.ts [hunk 18]
```diff
@@ -21,7 +24,8 @@ export class ParserPosition {
     return this.num >= 0 && this.num < this.lexerTokens.length;
   }
   public getRangeToEndLine() {
-    return this.lexerTokens[this.num]?.range.copyExtendEndOfLine();
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return this.lexerTokens[this.num]!.range.copyExtendEndOfLine();
```
Comment (Nik-Sch) [inline lib/parser/parser-base.ts:28]: throw a parserError if out of range

File: lib/parser/statement-body-parser.ts [hunk 19]
```diff
@@ -33,13 +33,13 @@ export class StatementBodyParser extends ParserBase {
       if (this.getToken().getLText() === '(') {
         const startRange = this.getToken().range;
         this.consumeToken(); // consume '('
-        (statementBody as OBlock).guardCondition = new ExpressionParser(this.state, statementBody, this.advanceClosingParenthesis()).parse();
+        (statementBody ).guardCondition = new ExpressionParser(this.state, statementBody, this.advanceClosingParenthesis()).parse();
```
Comment (Nik-Sch) [inline lib/parser/statement-body-parser.ts:36]:
```suggestion
        statementBody.guardCondition = new ExpressionParser(this.state, statementBody, this.advanceClosingParenthesis()).parse();
```

File: lib/vhdl-linter.ts [hunk 20]
```diff
@@ -161,40 +163,35 @@ export class VhdlLinter {
...
+      }
+      // await this.removeBrokenActuals();
+      if (profiling) {
+        console.log(`check ${i++}: ${Date.now() - start}ms`);
+        start = Date.now();
+      }
+      for (const checkerClass of rules) {
+        const checker = new checkerClass(this);
+        checker.check();
         if (profiling) {
```
Comment (Nik-Sch) [inline lib/vhdl-linter.ts:181]: this profiling might be misleading when an async operation is running...

**Review by Nik-Sch — CHANGES_REQUESTED** [review]
hated this pr

File: lib/parser/concurrent-statement-parser.ts [hunk 21]
```diff
@@ -58,8 +58,10 @@ export class ConcurrentStatementParser extends ParserBase {
       }
 
       const startI = this.state.pos.i;
-      const [constantName] = this.advancePast('in');
-
+      const [[constantName], inToken] = this.advanceParenthesisAware(['in'], true, true);
```
[the remark this answers is not in the document]
Comment (g0t00) [reply lib/parser/concurrent-statement-parser.ts:61]: I needed the inToken for the ParserError that is new in this commit.

File: package.json [hunk 22]
```diff
@@ -226,7 +226,6 @@
 		}
 	},
 	"activationEvents": [
```
[the remark this answers is not in the document]
Comment (g0t00) [reply package.json:228]: VScode was complaining that it is deprecated and auto generated from some other package.json attribute now.

File: lib/vhdl-linter.ts (hunk 20 above)
Comment (g0t00) [reply lib/vhdl-linter.ts:181]: I dont understand your point?

File: lib/languageFeatures/documentFormatting.ts [hunk 23]
```diff
@@ -53,12 +54,12 @@ export async function handleDocumentFormatting(params: DocumentFormattingParams,
   try {
     await new Promise<void>((resolve, reject) => {
       const args = ['-c', `emacs --batch --eval "(setq-default vhdl-basic-offset ${numSpaces})" ` +
-        `--eval "(setq load-path (cons (expand-file-name \\"${emacsLoadPath}\\") load-path))" ` +
-        ` -l ${emacsScripts} -f vhdl-batch-indent-region ${tmpFile}`];
+        `--eval "(setq load-path (cons (expand-file-name \\"${fileURLToPath(emacsLoadPath)}\\") load-path))" ` +
+        ` -l ${fileURLToPath(emacsScripts)} -f vhdl-batch-indent-region ${tmpFile}`];
```
[the remark this answers is not in the document]
Comment (g0t00) [reply lib/languageFeatures/documentFormatting.ts]: Checked and fixed formatter under win

File: lib/parser/interfaces.ts (hunk 1 above)
Comment (g0t00) [reply lib/parser/interfaces.ts:15]: It actually makes more sense with Partial though.... If it is not partial the check does not make sense...

File: lib/parser/objects.ts (hunk 3 above)
Comment (g0t00) [reply lib/parser/objects.ts:9]: Then it throws an error, what else do you want to do?

File: lib/parser/parser-base.ts (hunk 18 above)
Comment (g0t00) [reply lib/parser/parser-base.ts:28]: Based on which token? you need a range for that...

Comment (Nik-Sch) [reply lib/parser/parser-base.ts:26]: The last line of the file because we are out of range maybe?

File: lib/vhdl-linter.ts (hunk 20 above)
Comment (Nik-Sch) [reply lib/vhdl-linter.ts:180]: Some of the check functions are async and the promise is not awaited on.

File: package.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (Nik-Sch) [reply package.json:228]: Ah, ok, then just delete the whole array definition

File: lib/vhdl-linter.ts (hunk 20 above)
Comment (g0t00) [reply lib/vhdl-linter.ts:180]: Which ones? that should not be possible anymore?

Comment (Nik-Sch) [reply lib/vhdl-linter.ts:180]: The following:
`lib/rules/library.ts`
`lib/rules/not-declared.ts`
`lib/rules/port-declaration.ts`
`lib/rules/type-resolved.ts`
`lib/rules/unused.ts`
